### ISSUE TITLE
Use False/True instead of 0/1 in twisted/internet/defer.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4538]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4538|
|Type|enhancement|
|Created|2010-07-01 19:21:03Z|
|Branch|https://github.com/twisted/twisted/tree/booleans-in-defer-4538|

Booleans are great, I suppose.  Clarity of intent, or something like that.  `Deferred.called`, `Deferred._suppressAlreadyCalled`, `DeferredList.fireOnOne{Callback,Errback}, `DeferredLock.locked`... these aren't integers on which to perform arithmetic.  They're flags indicating one of two possible states.

(There seems to be some desire to do this in the [#4593](https://github.com/twisted/twisted/issues/4593) branch; so to shorten the diff there, I'm moving those changes here.)

<details><summary>Searchable metadata</summary>

```
trac-id__4538 4538
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_booleans_in_defer_4538 branches/booleans-in-defer-4538
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1278012063000000 1278012063000000
changetime__1278297626000000 1278297626000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [29561]) Branching to 'booleans-in-defer-4538'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Fairly straightforward.  [Documentation builder result](http://buildbot.twistedmatrix.com/builders/documentation/builds/581)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli set owner to @exarkun|
|-|-|

In defer.gatherResults, DeferredList is still initialized with `fireOnOneErrback = 1` instead of with a boolean.  Since this is in the defer module itself, it should probably be changed to be `fireOnOneErrback = True`.

However, should the rest of the code in Twisted be modified to use booleans when using deferreds?  Current usage is inconsistent - sometimes booleans are used, sometimes 0/1 flags.  For instance, twisted.conch.scripts.cftp uses the flags, but twisted.conch.test.test_transport uses booleans.   0/1 flags will still work (all tests pass, and [build results are fine](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/booleans-in-defer-4538)).  Perhaps if that is not in the scope of this branch a new ticket should be opened to change usage of flags in deferreds to booleans.

Other than these two points, looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [29580]) Change gatherResults to use False and True instead of 0 and 1

refs [#4538](https://github.com/twisted/twisted/issues/4538)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `assigned`|
|-|-|

> In defer.gatherResults, DeferredList is still initialized with fireOnOneErrback = 1

Fixed in r29580, along with an =0 in nearby code.

> However, should the rest of the code in Twisted be modified to use booleans when using deferreds? 

Not now.  This change makes sense because of [#4593](https://github.com/twisted/twisted/issues/4593) (or at least I'm assuming it does, since another developer made these changes in the branch for that ticket).  I've split them out to make the job of reviewing [#4593](https://github.com/twisted/twisted/issues/4593) easier.  This shouldn't be taken as a sign that we should go off to the rest of Twisted and start changing 1s to Trues and so on.

Where such a change will make maintenance easier, great.  But there's no point in doing it just to do it.

Also, there's no compatibility issue with code in Twisted continuing to pass 1 into these APIs rather than True.  The values are almost completely interchangable, and their differences (such as the result of using str() or repr() on them) won't make any difference to this code.

Thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [29581]) Merge booleans-in-defer-4538 

Author: therve, exarkun
Reviewer: cyli
Fixes: [#4538](https://github.com/twisted/twisted/issues/4538)

Document some of the attributes of `Deferred` and switch over most (perhaps all)
of `defer.py` to using actual booleans where it makes sense to do so.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
